### PR TITLE
Wip/hughsie/g main context push thread default take 2

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [X] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
@pwithnall this is right up your street; thanks again!

**Comment by hughsie** [comment]
Seems to fix the reported hang here: https://gitlab.gnome.org/GNOME/gnome-software/-/issues/1113#note_1000844 -- When Adam is awake I'll get him to try with those openQA instances.

File: libfwupd/fwupd-client.c [hunk 1]
```diff
@@ -121,13 +123,116 @@ fwupd_client_curl_helper_free (FwupdCurlHelper *helper)
...
+	gchar		*property_name;
+	guint		 signal_id;
+	FwupdDevice	*device;
+} FwupdClientContextHelper;
+
+static void
+fwupd_client_context_helper_free (FwupdClientContextHelper *helper)
+{
+	if (helper->device != NULL)
+		g_object_unref (helper->device);
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:137]: Could simplify this with `g_clear_object (&helper->device)`

File: libfwupd/fwupd-client.c [hunk 2]
```diff
@@ -121,13 +123,116 @@ fwupd_client_curl_helper_free (FwupdCurlHelper *helper)
...
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GMutexLocker) locker = NULL;
+
+	/* property */
+	if (helper->property_name != NULL)
+		g_object_notify (G_OBJECT (self), helper->property_name);
+
+	/* legacy signal name */
+	if (g_strcmp0 (helper->property_name, "status") == 0)
+		g_signal_emit (self, signals[SIGNAL_STATUS_CHANGED], 0, priv->status);
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:157]: Is reading `priv->status` here thread-safe?

File: libfwupd/fwupd-client.c [hunk 3]
```diff
@@ -121,13 +123,116 @@ fwupd_client_curl_helper_free (FwupdCurlHelper *helper)
...
+	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->idle_id_mutex);
+	g_autoptr(GSource) source = g_idle_source_new ();
+
+	/* run in the correct GMainContext and thread */
+	g_source_set_callback (source,
+			       fwupd_client_context_idle_cb,
+			       helper,
+			       (GDestroyNotify) fwupd_client_context_helper_free);
+	if (priv->idle_id != 0)
+		g_source_remove (priv->idle_id);
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:182]: This will cause the previously-pending signal emission to never happen, as the state from the `helper` object attached to the source being removed here is not stored anywhere else.

File: libfwupd/fwupd-client.c [hunk 4]
```diff
@@ -121,13 +123,116 @@ fwupd_client_curl_helper_free (FwupdCurlHelper *helper)
...
+fwupd_client_object_notify (FwupdClient *self, const gchar *property_name)
+{
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	FwupdClientContextHelper *helper = NULL;
+
+	/* shortcut */
+	if (priv->main_ctx == g_main_context_get_thread_default ()) {
+		g_object_notify (G_OBJECT (self), property_name);
+		if (g_strcmp0 (property_name, "status") == 0)
+			g_signal_emit (self, signals[SIGNAL_STATUS_CHANGED], 0, priv->status);
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:197]: Duplicating the code from `fwupd_client_context_idle_cb()` here means it’s possibly going to end up going out of sync.

File: libfwupd/fwupd-client.c [hunk 5]
```diff
@@ -305,19 +409,21 @@ fwupd_client_get_main_context (FwupdClient *self)
 /**
  * fwupd_client_set_main_context:
  * @self: A #FwupdClient
- * @main_ctx: #GMainContext
+ * @main_ctx: (nullable): #GMainContext or %NULL
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:412]: Would be good to clarify what `NULL` means.

File: libfwupd/fwupd-client.c [hunk 6]
```diff
@@ -491,6 +597,7 @@ fwupd_client_connect_get_proxy_cb (GObject *source,
 		fwupd_client_set_host_security_id (self, g_variant_get_string (val, NULL));
 
 	/* success */
+	fwupd_client_set_status (self, FWUPD_STATUS_IDLE);
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:600]: This change looks unrelated

File: libfwupd/fwupd-client.c [hunk 7]
```diff
@@ -2112,6 +2219,10 @@ fwupd_client_install_stream_async (FwupdClient *self,
  *
  * Install firmware onto a specific device.
  *
+ * NOTE: This method is thread-safe, but progress signals will be
+ * emitted in the default main context, if not explicitly set with
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:2223]: Would be good to clarify what you mean by ‘default main context’. Do you mean the global default main context (`g_main_context_default()`), the thread default main context (`g_main_context_get_thread_default()`) or something else specific to `FwupdClient`?

File: libfwupd/fwupd-context-test.c [hunk 8]
```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2020 Philip Withnall <philip@tecnocode.co.uk>
```
Comment (pwithnall) [inline libfwupd/fwupd-context-test.c:2]:
```suggestion
 * Copyright (C) 2020 Philip Withnall <pwithnall@endlessos.org>
```

(I did it on work time)

File: libfwupd/fwupd-client.c (hunk 1 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:137]: Fixed, ta.

File: libfwupd/fwupd-client.c (hunk 2 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:157]: Do you think it should be stored in the helper?

File: libfwupd/fwupd-client.c (hunk 3 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:182]: That was by design; is that right? My logic was that if nothing was "servicing" the GMainContext then it's pointless just queing them all up for when the method returns.

File: libfwupd/fwupd-client.c (hunk 4 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:197]: Maybe we add something like fwupd_client_set_status_safe() and do it there?

File: libfwupd/fwupd-client.c (hunk 6 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:600]: It's actually completely related; it causes a property change event which I'm using in the context test to check the GSource is working as we expect. I guess the proper fix is to do something like:

    val = g_dbus_proxy_get_cached_property (proxy, "Status");
    fwupd_client_set_status (self, g_variant_get_uint32 (val));

as I think that's set after the proxy is set up.

File: libfwupd/fwupd-client.c (hunk 7 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:2223]: > Do you mean the global default main context

Exactly that, will change to that wording.

File: libfwupd/fwupd-client.c (hunk 5 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:412]: Fixed, ta.

File: libfwupd/fwupd-client.c (hunk 6 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:600]: Yes, this works -- Status was missing completely!

File: libfwupd/fwupd-client.c (hunk 4 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:197]: This is what I did; better function name welcome.

File: libfwupd/fwupd-client.c (hunk 2 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:157]: The status is in reference to the daemon and shared state, rather than the transaction. All threads should have the same value and I think using a mutex for this legacy thing isn't required.

File: libfwupd/fwupd-client.c (hunk 3 above)
Comment (pwithnall) [reply libfwupd/fwupd-client.c:191]: I don’t think it’s right. It’s quite plausible for the `FwupdClient` thread to make two `notify` emissions in quick succession, fast enough that the main thread might not be able to process the first one before it’s removed and replaced with the second one.

If you want to do this properly, I’d suggest keeping a `GPtrArray` of the `GSource`s which are pending. Just store the pointers to the `GSource`s, don’t bother with the source IDs (they’re equivalent).

File: libfwupd/fwupd-client.c (hunk 6 above)
Comment (pwithnall) [reply libfwupd/fwupd-client.c]: Perhaps it should be a separate commit, with some explanation along the lines of what you’ve just said in the commit message.

File: libfwupd/fwupd-client.c (hunk 2 above)
Comment (pwithnall) [reply libfwupd/fwupd-client.c]: “All threads should have the same value” :rotating_light: Citation needed! :rotating_light: 

Storing it in the helper (so the value which is emitted in the signal is the value from when the signal was scheduled to be emitted) makes sense and should be simpler than reasoning about the thread-safety of `priv->status`.

File: libfwupd/fwupd-client.c [hunk 9]
```diff
@@ -121,13 +123,123 @@ fwupd_client_curl_helper_free (FwupdCurlHelper *helper)
...
+static void
+fwupd_client_context_helper_free (FwupdClientContextHelper *helper)
+{
+	g_clear_object (&helper->device);
+	g_object_unref (helper->self);
+	g_free (helper->property_name);
+	g_free (helper);
+}
+
+/* main_ctx == g_main_context_get_thread_default() */
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:142]: You could encode that as an assertion, for added safety:
```
g_return_if_fail (g_main_context_is_owner (priv->main_ctx));
```

Might be clearer to phrase the comment as `/* Always executed in the main context given by priv->main_ctx */`

File: libfwupd/fwupd-client.c (hunk 6 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c]: I did exactly that :)

File: libfwupd/fwupd-client.c [hunk 10]
```diff
@@ -121,13 +123,123 @@ fwupd_client_curl_helper_free (FwupdCurlHelper *helper)
...
+
+/* run callback in the correct thread */
+static void
+fwupd_client_object_notify (FwupdClient *self, const gchar *property_name)
+{
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	FwupdClientContextHelper *helper = NULL;
+
+	/* shortcut */
+	if (priv->main_ctx == g_main_context_get_thread_default ()) {
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:203]:
```suggestion
	if (g_main_context_is_owner (priv->main_ctx)) {
```
(See what `g_main_context_invoke()` does)

This is because of the difference between “what’s the default `GMainContext` in this thread at the moment” and “which `GMainContext` is currently being iterated in this thread”. One `GMainContext` might be the thread-default, but this function might actually be being called as part of a call chain from a `GSource` attached to a different (nested) `GMainContext` which is the one actually being iterated.

Put another way, one can think of `g_main_context_get_thread_default()` as being useful for deciding where to schedule something; and one can think of `g_main_context_is_owner()` as being useful for deciding where the currently-running code has been scheduled.

File: libfwupd/fwupd-client.c [hunk 11]
```diff
@@ -121,13 +123,123 @@ fwupd_client_curl_helper_free (FwupdCurlHelper *helper)
...
+
+/* run callback in the correct thread */
+static void
+fwupd_client_signal_emit_device (FwupdClient *self, guint signal_id, FwupdDevice *device)
+{
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	FwupdClientContextHelper *helper = NULL;
+
+	/* shortcut */
+	if (priv->main_ctx == g_main_context_get_thread_default ()) {
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:223]: Same as the comment above.

File: libfwupd/fwupd-client.c (hunk 6 above)
Comment (pwithnall) [reply libfwupd/fwupd-client.c]: Sorry, I think I was reviewing a stale version of the branch at that point! :partying_face:

File: libfwupd/fwupd-client.c (hunk 3 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:191]: > keeping a GPtrArray of the GSources which are pending

How would I remove the GSource after the idle source has run? Surely a better way would be to store the FwupdClientContextHelper's in an array as then you have something in fwupd_client_context_idle_cb() to remove. i.e. There's only one idle_id like now, but when fwupd_client_context_idle_cb is called we might emit multiple signals...

Comment (pwithnall) [reply libfwupd/fwupd-client.c:191]: Sure, that would work better

File: libfwupd/fwupd-client.c (hunk 9 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:142]: Agree, fixed, thanks.

File: libfwupd/fwupd-client.c (hunk 10 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:203]: TIL! Thanks, fixed.

File: libfwupd/fwupd-client.c (hunk 11 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:223]: Fixed.

File: libfwupd/fwupd-client.c (hunk 3 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:191]: > Sure, that would work better

**Narrator:** It did!

File: libfwupd/fwupd-client.c (hunk 2 above)
Comment (pwithnall) [reply libfwupd/fwupd-client.c:156]: Are you not going to store `priv->status` in the helper?

File: libfwupd/fwupd-client.c [hunk 12]
```diff
@@ -121,13 +124,127 @@ fwupd_client_curl_helper_free (FwupdCurlHelper *helper)
...
+fwupd_client_context_helper (FwupdClient *self, FwupdClientContextHelper *helper)
+{
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->idle_mutex);
+	g_autoptr(GSource) source = g_idle_source_new ();
+
+	/* run in the correct GMainContext and thread */
+	g_source_set_callback (source, fwupd_client_context_idle_cb, self, NULL);
+	if (priv->idle_id != 0)
+		g_source_remove (priv->idle_id);
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:195]: If there’s already a source attached to the context (i.e. `priv->idle_id != 0`) then you could just return early from this function, don’t remove the existing source, and don’t add a new one. Would save a bit of locking churn on that `GMainContext`

Comment (hughsie) [reply libfwupd/fwupd-client.c:195]: Agree. Something like this?

	FwupdClientPrivate *priv = GET_PRIVATE (self);
	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->idle_mutex);

	/* already a source attached to the context */
	if (priv->idle_id == 0) {
		g_autoptr(GSource) source = g_idle_source_new ();
		g_source_set_callback (source, fwupd_client_context_idle_cb, self, NULL);
		priv->idle_id = g_source_attach (g_steal_pointer (&source), priv->main_ctx);
	}

	/* run in the correct GMainContext and thread */
	g_ptr_array_add (priv->idle_sources, helper);

Comment (pwithnall) [reply libfwupd/fwupd-client.c]: Yes, although I think the `already a source attached to the context` comment needs inverting (a source isn’t attached if that branch is taken)

Comment (hughsie) [reply libfwupd/fwupd-client.c]: Ohh yes. indeed. thanks!

**Review by pwithnall — APPROVED** [review]
lgtm

Hopefully that’s the last of the fwupd main context issues and solves things for gnome-software. Thanks for taking the time to dig through it all :)

**Comment by hughsie** [comment]
> Thanks for taking the time to dig through it all :)

No, quite the opposite! Thank you for explaining it to me!
